[CORL-233] Create a user drawer with comment streams

nick-funk opened the pull request:
## What does this PR do?

Adds a user drawer to the moderation area when you click on a moderation card username.

## How do I test this PR?

- Go to moderation area
- Find a moderation card
- Click on the user name
- See the user drawer appear with user details and comment streams
- Should have an approve and a reject comment stream

nick-funk committed:
Create preliminary user history drawer with user fragment

CORL-233

nick-funk committed:
Create preliminary user history drawer with user fragment

CORL-233

nick-funk committed:
Create preliminary user history drawer with user fragment

CORL-233

nick-funk committed:
Set font size for the user details to match spec at 16 px

CORL-233

nick-funk committed:
Set font size for the user details to match spec at 16 px

CORL-233

nick-funk committed:
Set font size for the user details to match spec at 16 px

CORL-233

nick-funk committed:
Update username font style/size, add horizontal rule for user history drawer

CORL-233

nick-funk committed:
Update username font style/size, add horizontal rule for user history drawer

CORL-233

nick-funk committed:
Update username font style/size, add horizontal rule for user history drawer

CORL-233

nick-funk committed:
Set user detail min width to align the copy buttons in user history drawer

CORL-233

nick-funk committed:
Set user detail min width to align the copy buttons in user history drawer

CORL-233

nick-funk committed:
Set user detail min width to align the copy buttons in user history drawer

CORL-233

nick-funk committed:
Set min width of the user history drawer to match spec

CORL-233

nick-funk committed:
Set min width of the user history drawer to match spec

CORL-233

nick-funk committed:
Set min width of the user history drawer to match spec

CORL-233

nick-funk committed:
Remove min width for detail area and card to allow responsive sizing

Allows the copy buttons to sit just to the right of the user
details text.

CORL-233

nick-funk committed:
Remove min width for detail area and card to allow responsive sizing

Allows the copy buttons to sit just to the right of the user
details text.

CORL-233

nick-funk committed:
Remove min width for detail area and card to allow responsive sizing

Allows the copy buttons to sit just to the right of the user
details text.

CORL-233

nick-funk committed:
Update snapshots to handle clickable username in moderation comment cards

CORL-233

nick-funk committed:
Update snapshots to handle clickable username in moderation comment cards

CORL-233

nick-funk committed:
Update snapshots to handle clickable username in moderation comment cards

CORL-233

nick-funk committed:
Create initial pagination of all comments in user history drawer

Shows all comments with a rubbish "load more" button as we haven't
decided if we want to do a load more button or infinite scrolling
solution.

Flex layout helps it create a scroll bar when needed to traverse through
the comment stream.

CORL-233

nick-funk committed:
Let the user history drawer fill full height of page

CORL-233

nick-funk committed:
Let the user history drawer fill full height of page

CORL-233

nick-funk committed:
Add a close button to the user history drawer

CORL-233

nick-funk committed:
Add a close button to the user history drawer

CORL-233

nick-funk committed:
Merge remote-tracking branch 'origin/feature/CORL-233' into feature/CORL-233-queue

# Conflicts:
#	src/core/client/admin/routes/Moderate/ModerateCard/ModerateCardContainer.tsx
#	src/core/client/admin/routes/Moderate/UserHistoryDrawer/UserHistoryDrawerContainer.css
#	src/core/client/admin/routes/Moderate/UserHistoryDrawer/UserHistoryDrawerContainer.tsx

nick-funk committed:
Set the user history drawer to fill full height of window

CORL-233

nick-funk committed:
Convert ModerateCardContainer to a FunctionComponent

CORL-233

nick-funk committed:
Preliminarily get all and rejected comments showing in the user drawer

CORL-233

nick-funk committed:
Merge remote-tracking branch 'origin/feature/CORL-233' into feature/CORL-233-queue

nick-funk committed:
Show spinner while all comments are loading in user history drawer

CORL-233

nick-funk committed:
Style tabs within the user history drawer to be secondary

CORL-233

nick-funk committed:
Center load more button in the comments for the user history drawer

CORL-233

nick-funk committed:
Fix scroll bars on nested flexboxes in the user history drawer

CORL-233

nick-funk committed:
Tweak styling for the user history drawer comments section

Add margins where necessary.
Style the Load More button to match the comment feed elsewhere.

CORL-233

nick-funk committed:
Fix background of copy buttons to match the user history drawer

CORL-233

nick-funk committed:
Preliminarily get rejected comments showing in the user history tabs

Still need to consolidate any duplication between the all and rejected
comments containers/queries, but this works correctly as is.

CORL-233

nick-funk committed:
Show hover and active background color on moderate card usernames

CORL-233

nick-funk committed:
Hide load more buttons on comment streams when no comments available

CORL-233

nick-funk committed:
Handle when no comments are available in user history drawer streams

CORL-233

nick-funk committed:
Actually show all comments in the user history all comments tab

Previously was filtering out rejected and other pre-mod
status-ed comments.

CORL-233

nick-funk committed:
Style the user history drawer cards to match the spec

CORL-233

nick-funk committed:
Add localizations around user history drawer tabs

CORL-233

nick-funk committed:
Allow a mini mode for moderate cards that are shown in the user drawer

CORL-233

nick-funk committed:
Style user drawer comment tabs to match the spec

CORL-233

nick-funk committed:
Update snapshots to account for mini moderate cards

Updates the fact that a separator on regular sized cards
now has an additional class name to turn on the divider's
visible border.

CORL-233

wyattjoh commented on src/core/client/admin/routes/Moderate/ModerateCard/ModerateCardContainer.tsx:
```diff
@@ -54,67 +57,83 @@ function isFeatured(comment: ModerateCardContainer_comment) {
+  rejectComment,
+  featureComment,
+  unfeatureComment,
+  mini,
```
```suggestion
  mini = false,
```

wyattjoh commented on src/core/client/admin/routes/Moderate/ModerateCard/ModerateCardContainer.tsx:
```diff
@@ -142,19 +162,28 @@ class ModerateCardContainer extends React.Component<Props> {
           storyHref={getModerationLink("default", comment.story.id)}
-          onModerateStory={this.handleModerateStory}
+          onModerateStory={handleModerateStory}
+          mini={mini === undefined ? false : mini}
```
Remove in favour of above suggestion.

wyattjoh commented on src/core/client/admin/routes/Moderate/ModerateCard/ModerateCardContainer.tsx:
```diff
@@ -142,19 +162,28 @@ class ModerateCardContainer extends React.Component<Props> {
-          onModerateStory={this.handleModerateStory}
+          onModerateStory={handleModerateStory}
+          mini={mini === undefined ? false : mini}
+          showUsername={showUsername === undefined ? true : showUsername}
```
Remove in favour of above suggestion.

wyattjoh commented on src/core/client/admin/routes/Moderate/ModerateCard/ModerateCardContainer.tsx:
```diff
@@ -54,67 +57,83 @@ function isFeatured(comment: ModerateCardContainer_comment) {
+  featureComment,
+  unfeatureComment,
+  mini,
+  showUsername,
```
```suggestion
  showUsername = true,
```

wyattjoh commented on src/core/client/admin/routes/Moderate/UserHistoryDrawer/UserHistoryDrawerContainer.tsx:
```diff
@@ -0,0 +1,113 @@
+  Modal,
+  Typography,
+} from "coral-ui/components";
+
```
Import ordering

wyattjoh commented on src/core/client/admin/routes/Moderate/UserHistoryDrawer/UserHistoryRejectedComments.tsx:
```diff
@@ -0,0 +1,149 @@
+import React, { FunctionComponent } from "react";
+import { graphql, RelayPaginationProp } from "react-relay";
+
+import { Button } from "coral-ui/components/";
```
Import ordering, and extra `/` in import path.

wyattjoh commented on src/core/client/admin/routes/Moderate/UserHistoryDrawer/UserHistoryRejectedComments.tsx:
```diff
@@ -0,0 +1,149 @@
+    ? props.user.rejectedComments.edges.map(edge => edge.node)
+    : [];
+
+  const [loadMore, isLoadingMore] = useLoadMore(props.relay, 5);
```
Hooks cannot be used after non-hook/conditional code is ran, re-order.

wyattjoh commented on src/core/client/admin/routes/Moderate/UserHistoryDrawer/UserHistoryRejectedComments.tsx:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    loadMore();
+  };
```
Use `useCallback`.

wyattjoh commented on src/core/client/admin/routes/Moderate/UserHistoryDrawer/UserHistoryRejectedComments.tsx:
```diff
@@ -0,0 +1,149 @@
+    return (
+      <div className={styles.error}>{`${
+        props.user.username
+      } does not have any rejected comments.`}</div>
```
Must be wrapped in translation.

wyattjoh replied:
Replace error class with the `<Callout />` component from `ui`.

wyattjoh commented on src/core/client/admin/routes/Moderate/UserHistoryDrawer/UserHistoryTabs.tsx:
```diff
@@ -0,0 +1,83 @@
+  const [currentTab, setCurrentTab] = useState("ALL");
+
+  const onTabChanged = useCallback(
+    (tab: any) => {
```
Should be a type.

```ts
type Tabs = "ALL" | "REJECTED";
// ...
const [currentTab, setCurrentTab] = useState<Tabs>("ALL");
// ...
const onTabChanged = useCallback(
  (tab: Tabs) => {
```

wyattjoh commented on src/core/client/admin/routes/Moderate/UserHistoryDrawer/UserHistoryTabs.tsx:
```diff
@@ -0,0 +1,83 @@
+        onTabClick={onTabChanged}
+        className={styles.tabBar}
+      >
+        <Tab tabID={"ALL"} onTabClick={onTabChanged}>
```
```suggestion
        <Tab tabID="ALL" onTabClick={onTabChanged}>
```

wyattjoh commented on src/core/client/admin/routes/Configure/HorizontalRule.tsx:
```diff
@@ -1,9 +1,15 @@
+}
+
+const HorizontalRule: FunctionComponent<Props> = ({ children, className }) => (
+  <hr className={cn(styles.root, className)} />
```
Do not use this `<HorizontalRule />` component inside the `Moderate` route as this is from another route. Replace usage of `<HorizontalRule />` with an `<hr />` or a new `<HorizontalRule />` for the `Moderate` route.

wyattjoh commented on src/core/client/admin/routes/Moderate/UserHistoryDrawer/UserHistoryRejectedComments.tsx:
```diff
@@ -0,0 +1,149 @@
+            mini
+            showUsername={false}
+          />
+          {i !== comments.length - 1 && <HorizontalRule />}
```
No magiccccc

nick-funk committed:
Add user drawer to the queue instead of childed to the moderate card

Prevents the user drawer from prematurely closing when we approve
all items in the moderate queue.

CORL-233

nick-funk committed:
Update snapshots to account for user drawer on queues

CORL-233

wyattjoh commented on src/core/client/admin/routes/Moderate/ModerateCard/ModerateCardContainer.tsx:
```diff
@@ -40,6 +39,7 @@ interface Props {
   showStoryInfo: boolean;
   mini?: boolean;
   showUsername?: boolean;
+  usernameClicked?: (userID: string) => void;
```
Suggest changing to `onUsernameClicked` to stay consistent with method props.

wyattjoh commented on src/core/client/admin/routes/Moderate/UserHistoryDrawer/UserHistoryDrawerContainer.tsx:
```diff
@@ -21,93 +23,120 @@ import styles from "./UserHistoryDrawerContainer.css";
-            <div className={styles.comments}>
-              <UserHistoryTabs userID={user.id} />
-            </div>
+          {open && userID && (
```
The modal will always show the contents when the `Modal` `open` prop is true, so move this instead so that:

```tsx
<Modal open={open && userID} // ...
```

And remove the:

```tsx
{open && userID && // ...
```

wyattjoh commented on src/core/client/admin/routes/Moderate/UserHistoryDrawer/UserHistoryDrawerContainer.tsx:
```diff
@@ -21,93 +23,120 @@ import styles from "./UserHistoryDrawerContainer.css";
-              <UserHistoryTabs userID={user.id} />
-            </div>
+          {open && userID && (
+            <QueryRenderer<QueryTypes>
```
Suggest moving this into another component `UserHistoryDrawerQuery`.

nick-funk committed:
Set mini and username defaults more cleanly on ModerateCardContainer

CORL-233

nick-funk committed:
Rename usernameClicked to onUsernameClicked

CORL-233

nick-funk committed:
Use callback on loadMore calls within user drawer queries

CORL-233

nick-funk committed:
Localize the no comment text for the user drawer comment streams

CORL-233

nick-funk committed:
Localize the user not found callout in the user drawer comment streams

CORL-233

nick-funk committed:
Remove superfluous curly braces

CORL-233

nick-funk committed:
Extract user history drawer internals into a distinct query component

CORL-233

nick-funk committed:
Move comments conditional below useCallback initializations

CORL-233

wyattjoh committed:
feat: fixed consistency issues

nick-funk committed:
Use a concrete tab type on user drawer comment stream tabs

CORL-233

nick-funk committed:
Use <hr/> instead of <HorizontalRule>

CORL-233

nick-funk committed:
Clarify logic for showing horizontal rule after comments in user drawer

CORL-233

nick-funk committed:
Remover bottom border on user drawer tab bar

CORL-233
